src/protocol.rs

```diff
@@ -121,145 +91,56 @@ impl Participant {
+                message.id(),
+                self.sid
+            );
+            Err(CallerError::WrongSessionId)?
         }
-    }
 
-    /// Produces a message to signal to this participant that auxinfo generation
-    /// is ready for the specified identifier.
-    #[instrument(skip_all)]
-    pub fn initialize_auxinfo_message(&self, auxinfo_identifier: Identifier) -> Message {
-        info!("Auxinfo generation is ready.");
-        Message::new(
-            MessageType::Auxinfo(AuxinfoMessageType::Ready),
-            auxinfo_identifier,
-            self.id,
-            self.id,
-            &[],
-        )
-    }
+        // Check that message belongs to correct protocol
+        match (message.message_type(), P::protocol_type()) {
+            (MessageType::Auxinfo(_), ProtocolType::AuxInfo)
+            | (MessageType::Keygen(_), ProtocolType::Keygen)
+            | (MessageType::Presign(_), ProtocolType::Presign) => {}
+            _ => {
+                error!(
+                    "Message type did not match type of this participant: got {:?}, expected {:?}",
+                    message.message_type(),
+                    P::protocol_type()
+                );
+                Err(CallerError::WrongProtocol)?
+            }
+        };
 
-    /// Produces a message to signal to this participant that keyshare
-    /// generation is ready for the specified identifier
-    #[instrument(skip_all)]
-    pub fn initialize_keygen_message(&self, keygen_identifier: Identifier) -> Message {
-        info!("Keyshare generation is ready.");
-        Message::new(
-            MessageType::Keygen(KeygenMessageType::Ready),
-            keygen_identifier,
-            self.id,
-            self.id,
-            &[],
-        )
+        // Handle it!
+        let outcome = self
+            .participant
+            .process_message(rng, message, &self.input)?;
+        let (output, messages) = outcome.into_parts();
+        Ok((message.id(), output, messages))
     }
 
-    /// Produces a message to signal to this participant that presignature
-    /// generation is ready for the specified identifier. This also requires
-    /// supplying the associated auxinfo identifier and keyshare identifier.
-    /// `auxinfo_identifier`, `keyshare_identifier` and `identifier` correspond
-    /// to session identifiers.
+    /// Produce a message to signal to this participant that the protocol can
+    /// begin.
     #[instrument(skip_all)]
-    pub fn initialize_presign_message(
-        &mut self,
-        auxinfo_identifier: Identifier,
-        keyshare_identifier: Identifier,
-        identifier: Identifier,
-    ) -> Result<Message> {
-        info!("Presignature generation is ready.");
-        self.presign_participant.initialize_presign_message(
-            auxinfo_identifier,
-            keyshare_identifier,
-            identifier,
-        )
+    pub fn initialize_message(&self) -> Message {
```

Comment:
note: this replaces three functions (`initialize_{keygen | auxinfo | presign}_message`).